File: JAVA/Hindex.java
/*
programmers 정렬 H-Index
https://programmers.co.kr/learn/courses/30/parts/12198
*/


import java.util.Arrays;

class Hindex_Solution {
    public int solution(int[] citations) {
        // 과학자가 발표한 논문의 수 오름차순 정렬
        Arrays.sort(citations);
        
        int leng = citations.length;
        
        int answer =0;
        for(int i=leng-1;i>=0;i--)
            if(i+1 <= citations[leng-i-1]){
                answer = i+1;
                break;
            }
        
        return answer;
    }
}

public class Hindex {
    public static void main(String[] args){
        // Test Case 1
        // int[] citations = {3, 0, 6, 1, 5};

        // My Test Case 1
        int[] citations = {8, 9, 5, 6, 0, 2, 10};

        // My Test Case 2
        // int[] citations = {5, 5, 5, 5};

        Hindex_Solution hindex = new Hindex_Solution();
        int answer = hindex.solution(citations);

        System.out.println(answer);
    }
}

File: Python/Sort/H_index.py
"""
https://programmers.co.kr/learn/courses/30/parts/12198
PROGRAMMERS SORT H INDEX
"""

def solution(citations):
    answer = 0

    citations.sort()
    count = 0
    for i in range(len(citations)-1, -1, -1):
        if i+1 <= citations[len(citations)-1-i]: count+=1
            
    return count

def main():
    citations1 = [3, 0, 6, 1, 5]
    citations2 = [0, 2, 5, 6, 8, 9, 10]
    citations3 = [5, 5, 5, 5, 5]
    citations4 = [5, 5, 5, 5]
    citations5 = [31, 66]

    result1 = 3
    result2 = 5
    result3 = 5
    result4 = 4
    result5 = 2

    if solution(citations1) == result1: print("RIGHT1")
    if solution(citations2) == result2: print("RIGHT2")
    if solution(citations3) == result3: print("RIGHT3")
    if solution(citations4) == result4: print("RIGHT4")
    if solution(citations5) == result5: print("RIGHT5")

if __name__=="__main__":
    main()
    